### ISSUE TITLE
Option to see/log raw NetworkClient data

### ISSUE THREAD

**Issue by large** [cosullivan/SmtpServer#118]
Hi @cosullivan , thank you for a awesome package, I have been using it for months and it is stable and fine.

In some occasions I want to know all the raw data received/sent between server and client.
Your "LogVerbose" should be implemented for such a thing, LogRawNetworkData()?

Or are there any other places I could find what I'm looking for?
The TracingSmtpCommandVisitor is to "clean" for tracing errors/challenges by the spammers.

**Comment by cosullivan** [comment]
Hi @large,

Thanks for your feedback.

I don't think adding a method to the Logger is the right place for this, but there is a way in which you can take control of the underlying stream that is being used and in which case you could log all of the read and write operations there. 

The way to do this is to create your own IEndpointListenerFactory and if you look through that you will be able to create your own IEndpointListener instances that return the underlying stream.

You should be able to subclass/delegate the existing implementations to do what you want. 

If you think that approach will work (ie, having access to the underlying stream) then I can create a sample if you get stuck.

Thanks, Cain.

**Comment by cosullivan** [comment]
FYI, I've added a sample here;
https://github.com/cosullivan/SmtpServer/blob/master/Src/SampleApp/Examples/CustomEndpointListenerExample.cs

**Comment by large** [comment]
@cosullivan this is fantastic!
I did not know about this feature (and I have to work on my C# skills :))

Works as a charm and makes it possible to count connections, see violation etc...
Thank you very much!

**Comment by large** [comment]
For testing purposes I added an override on Close();

```
public override void Close()
{
     Console.WriteLine("Streambuffer: " + Streambuffer);
}

```
On unsecure connections: Close() is called
On secure connections: Close() is never called (but several flushes...) data is also unreadable...

After checking the code a secure stream is SslStream() not Stream().
So is the cleanest solution to add another custom SslStream() and override UpgradeAsync(...)?
Or do you have other tricks in your sleeve? :)

**Comment by cosullivan** [comment]
Ok, yeah that makes sense that you wont be able to read the stream contents as it will be encrypted. 

There's no way with the current code then to do what you want. The two most likely changes would be to create some sort of IStreamFactory that you could implement that managed the stream creation. 

The second is whether I expose the creation of INetworkClient such that you could create you own implementation of that. 

I will look into these options in the next few days.

**Comment by cosullivan** [comment]
FYI I am working on a solution and you can follow the progress here;
https://github.com/cosullivan/SmtpServer/tree/feature/DelegatingStreamAccess

Thanks, Cain.

**Comment by large** [comment]
@cosullivan that is awesome! Looking forward to try the end-result

**Comment by large** [comment]
@cosullivan did you get this thing working?

**Comment by cosullivan** [comment]
@large I'm about half way through, but I've now got some spare time over the next few months that I am going to be working on a few different things on this project and going back to this is my first item.

**Comment by cosullivan** [comment]
Hi @large,

If you are interested I have a working sample now in a branch here;
https://github.com/cosullivan/SmtpServer/blob/feature/DelegatingStreamAccess/Src/SampleApp/Examples/CustomEndpointListenerExample.cs

I am going to make a few more changes to the API before merging as a new version, but those changes are separate to this.

Thanks, Cain.

**Comment by cosullivan** [comment]
This is now released in v7.0.0.
